ledger/src/bank_forks_utils.rs

```diff
@@ -222,77 +222,101 @@ fn bank_forks_from_snapshot(
         process::exit(1);
     }
 
-    let bank = match process_options.use_snapshot_archives_at_startup {
-        UseSnapshotArchivesAtStartup::Always => {
-            // Given that we are going to boot from an archive, the append vecs held in the snapshot dirs for fast-boot should
-            // be released.  They will be released by the account_background_service anyway.  But in the case of the account_paths
-            // using memory-mounted file system, they are not released early enough to give space for the new append-vecs from
-            // the archives, causing the out-of-memory problem.  So, purge the snapshot dirs upfront before loading from the archive.
-            snapshot_utils::purge_old_bank_snapshots(&snapshot_config.bank_snapshots_dir, 0, None);
-
-            let (bank, _) = snapshot_utils::bank_from_snapshot_archives(
-                &account_paths,
-                &snapshot_config.bank_snapshots_dir,
-                &full_snapshot_archive_info,
-                incremental_snapshot_archive_info.as_ref(),
-                genesis_config,
-                &process_options.runtime_config,
-                process_options.debug_keys.clone(),
-                None,
-                process_options.account_indexes.clone(),
-                process_options.limit_load_slot_count_from_snapshot,
-                process_options.shrink_ratio,
-                process_options.accounts_db_test_hash_calculation,
-                process_options.accounts_db_skip_shrink,
-                process_options.verify_index,
-                process_options.accounts_db_config.clone(),
-                accounts_update_notifier,
-                exit,
-            )
-            .expect("load bank from snapshot archives");
+    let latest_snapshot_archive_slot = std::cmp::max(
+        full_snapshot_archive_info.slot(),
+        incremental_snapshot_archive_info
+            .as_ref()
+            .map(SnapshotArchiveInfoGetter::slot)
+            .unwrap_or(0),
+    );
+    let latest_bank_snapshot =
+        snapshot_utils::get_highest_bank_snapshot_post(&snapshot_config.bank_snapshots_dir);
 
-            bank
+    let will_startup_from_snapshot_archives = match process_options.use_snapshot_archives_at_startup
+    {
+        UseSnapshotArchivesAtStartup::Always => true,
+        UseSnapshotArchivesAtStartup::Never => false,
+        UseSnapshotArchivesAtStartup::WhenNewest => {
+            Some(latest_snapshot_archive_slot)
+                > latest_bank_snapshot
+                    .as_ref()
+                    .map(|bank_snapshot| bank_snapshot.slot)
```

Comment:
Maybe my brain is wrong, but I **typically** don't like `Option` comparisons unless necessary, because it's not always obvious (at least to me) what it does when something is `None`.

I think in this case we can be a bit more explicit about what happens by doing something like:
```suggestion
        UseSnapshotArchivesAtStartup::WhenNewest => lastest_bank_snapshot
            .as_ref()
            .map(|bank_snapshot| latest_snapshot_archive_slot > bank_snapshot.slot)
            .unwrap_or(true),
```

Reply:
I thought about this one, which I think is possibly(?) subtlely different. If we have a snapshot archive at slot 0 this would now be false, even if we don't have a `latest_bank_snapshot`.

Obviously, we don't need to worry about that for any running networks, but not sure if there's potentially an edge case for local cluster tests or something.

I can get trying to keep the comparison at the top, I guess I was thinking of a more concise version of:
```rust
if let Some(latest_bank_snapshot) = latest_bank_snapshot {
   latest_snapshot_archive_slot > latest_bank_snapshot.slot
} else {
   true
}
```
i.e. we check first if we even have a `latest_bank_snapshot`, then do the comparison.

I think I'll leave it up to you, just comment here and re-request review when it's how you want it 😄

Reply:
9ab6ae7: "lastest" 😆 I checked the formatting, but not the spelling